QDB-3317 Pinned columns

solatis opened the pull request.

solatis committed:
Reproduce bottleneck

solatis committed:
Add 'pinned' writer placeholder

solatis committed:
Parametrize writer type in benchmarks

solatis committed:
Adds functionality to look up shard size

We need to know the shard size for the pinned columns writer

solatis committed:
Add a few utility functions

solatis committed:
Doubles are fully working for the pinned writer

solatis committed:
Templatized column pinner

solatis committed:
Add support for pinned writing of int64 colukmns

solatis committed:
Restructure pinned column functions into separate file

solatis committed:
Timestamps work with pinned columns

solatis committed:
[WiP] Blob support

solatis committed:
Add missing file

solatis committed:
Blobs work

solatis committed:
String support fully implemented

solatis committed:
Implement 'extra tables' support

solatis committed:
Explicitly refer to jlong/qdb_time_t when appropriate

solatis committed:
Remove debug logs

solatis committed:
Don't run benchmarks by default

solatis committed:
Fix performance trace tests to account for additional call

We now lookup shard sizes in the Table's constructor, which implies
one additional lookup, which implies one additional performance trace
entry.

solatis committed:
Fix javadoc issue

solatis committed:
Better timespec toString

solatis committed:
Remove debug

solatis committed:
Use jlong instead of long

jlong is longer than long on windows

solatis committed:
Move some of the column pinner functions to a standalone cpp file

solatis committed:
Add pinned columns to cmakelist

solatis committed:
Clean up some code

solatis committed:
qdb_int_t is actually jlong -> qdb_int_t

solatis committed:
Fix things (and stuff)

solatis committed:
Fix the fix

solatis committed:
Patch for the fix of the fix

solatis committed:
Move some includes to source files

solatis committed:
Dynamically resize arrays

solatis committed:
Fix comments

solatis committed:
Merge branch 'master' into pinned-columns

solatis committed:
Explicitly release memory after setting blob/string values

mkurdej commented on src/main/c++/net/quasardb/qdb/jni/byte_buffer.cpp:
```diff
@@ -1,3 +1,4 @@
+#include <iostream>
```
Why? Oh why are you killing innocent kittens?

solatis replied:
Did I add you to the reviewers list? I think not.

mkurdej commented on src/test/java/net/quasardb/qdb/ts/WriterExtraTablesBenchmarkTest.java:
```diff
@@ -0,0 +1,126 @@
+import net.quasardb.qdb.exception.InvalidIteratorException;
+import net.quasardb.qdb.exception.InvalidArgumentException;
+
+// @State(Scope.Benchmark)
```
Why is it all commented?

mkurdej commented on src/test/java/net/quasardb/qdb/ts/WriterAppendBenchmarkTest.java:
```diff
@@ -0,0 +1,106 @@
+import net.quasardb.qdb.exception.InvalidIteratorException;
+import net.quasardb.qdb.exception.InvalidArgumentException;
+
+// @State(Scope.Benchmark)
```
Again. Why is it all commented?

solatis replied:
I apologize, this was temporarily to have some benchmarks inside the unit test suite. It's already been refactored out as a separate component.

mkurdej commented on test.sh:
```diff
@@ -8,4 +8,6 @@ mkdir build || true && \
-    mvn test
+    #mvn  '-Dtest=WriterExtraTablesBenchmarkTest*' test
+    # mvn  '-Dtest=TableTest*' test
+    mvn '-Dtest=WriterTest#canInsertNullRows' test
```
Euuh, isn't it a change that shouldn't have been committed? It seems as it runs a single test only, nope?

solatis replied:
You are correct, this script isn't used anywhere in CI, just a convenience script for development. Will undo the change.

solatis committed:
Optimize pinned column matrix

We're now keeping explicit track of which columns actually have rows,
and not pin any columns that do not have rows. This has a significant
improvement with sparse datasets.

solatis committed:
Optimizations, refactorings

solatis committed:
Benchmark suite with jmh

solatis committed:
Convenience function to remove tables

solatis committed:
Move some things around

solatis committed:
These have been migrated

solatis committed:
Revert change

solatis committed:
Merge branch 'master' into pinned-columns

solatis committed:
Fix scope

solatis committed:
Fix indent

solatis committed:
Remove benchmarks from this PR
